exonum-node/src/state.rs

```diff
@@ -1097,21 +1096,29 @@ impl State {
     }
 
     /// Adds propose from this node to the proposes list for the current height. Such propose
-    /// cannot contain unknown transactions. Returns hash of the propose.
-    pub(super) fn add_self_propose(&mut self, msg: Verified<Propose>) -> Hash {
+    /// cannot contain unknown transactions. Returns the hash of the propose.
+    pub(super) fn add_self_propose(
+        &mut self,
+        msg: Verified<Propose>,
+        blockchain: &mut BlockchainMut,
+    ) -> Hash {
         debug_assert!(self.validator_state().is_some());
+
+        let round = msg.payload().round;
+        blockchain.persist_changes(
+            |schema| schema.save_message(round, msg.clone()),
+            "Cannot save `Propose` to message cache",
+        );
+
         let propose_hash = msg.object_hash();
         self.proposes.insert(
             propose_hash,
             ProposeState {
                 propose: msg,
                 unknown_txs: HashSet::new(),
                 block_hash: None,
-                // TODO: For the moment it's true because this code gets called immediately after
-                // saving a propose to the cache. Think about making this approach less error-prone.
-                // (ECR-1635)
                 is_saved: true,
-                // We expect ourself not to produce invalid proposes.
+                // We expect our node not to produce invalid proposes.
```

Comment:
```suggestion
                // We expect that our node not to produce invalid proposes.
```

Reply:
I think "expect [A] to [B]" does not require "that"; with "that", it would look like "expect that [A] [B]s".